exn.ml

```diff
@@ -0,0 +1,16 @@
+let [@inline never] print_endline x = print_endline x
+
+let g () = failwith "foo"
+
+let f () =
+  let r = ref "foo" in
+  let s = ref "bar" in
+  try
+    (g [@inlined never]) ()
```

Comment:
Maybe you could try to modify the references before raising, to check that the correct value gets picked up ?

Reply:
I'll turn this into a proper test case too, and add one involving async exns as discussed.